include/nbl/builtin/hlsl/emulated_float64_t.hlsl

```diff
@@ -0,0 +1,630 @@
+#include <nbl/builtin/hlsl/cpp_compat.hlsl>
+
+#ifdef __HLSL_VERSION
+#define LERP lerp
+#else
+#define LERP nbl::hlsl::lerp
+#endif
+
+#ifdef __HLSL_VERSION
+#define ABS abs
+#else
+#define ABS std::abs
+#endif
+
+// TODO: inline function
+#define EXCHANGE(a, b) \
+   do {                \
+       a ^= b;         \
+       b ^= a;         \
+       a ^= b;         \
+   } while (false)

```

Comment:
add a `nbl::hlsl::swap(NBL_REF_ARG,NBL_REF_ARG)` counterpart to `std::swap` in `algorithm.hlsl`

then you can specialize for integer types (if you implement via a struct in `impl` namespace) so that `^=` is used instead of a temporary variable